### PR TITLE
benchmark: add auto-compact flags

### PR DESCRIPTION
**Pull request by gyuho**
For https://github.com/coreos/etcd/issues/4633.

### CODE REVIEW

**Comment by xiang90** [comment]
can you please try to get some initial benchmark result out of this? So we know this is useful?

**Comment by gyuho** [comment]
Sure, I will rework on this with some initial results. Thanks!

**Comment by gyuho** [comment]
@xiang90 PTAL. Added `same-key` option:

Results:

```
# without compaction
benchmark --conns=10 --clients=1000 put --total=1000000 --same-key

Summary:
  Total:        96.4195 secs.
  Slowest:      0.2931 secs.
  Fastest:      0.0179 secs.
  Average:      0.0964 secs.
  Stddev:       0.0265 secs.
  Requests/sec: 10371.3499

Response time histogram:
  0.018 [1]     |
  0.045 [5966]  |
  0.073 [183154]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.100 [416139]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.128 [279370]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.156 [92690] |∎∎∎∎∎∎∎∎
  0.183 [17999] |∎
  0.211 [2784]  |
  0.238 [915]   |
  0.266 [550]   |
  0.293 [432]   |

Latency distribution:
  10% in 0.0650 secs.
  25% in 0.0777 secs.
  50% in 0.0935 secs.
  75% in 0.1120 secs.
  90% in 0.1308 secs.
  95% in 0.1431 secs.
  99% in 0.1687 secs.
```

```
# with compaction
benchmark --conns=10 --clients=1000 put --total=1000000 --same-key --compact-index-delta=1000 --compact-interval=0.5s

Summary:
  Total:        102.2650 secs.
  Slowest:      0.5117 secs.
  Fastest:      0.0194 secs.
  Average:      0.1022 secs.
  Stddev:       0.0283 secs.
  Requests/sec: 9779.9628

Response time histogram:
  0.019 [1]     |
  0.069 [76991] |∎∎∎∎
  0.118 [681804]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.167 [224576]        |∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.216 [15464] |
  0.266 [310]   |
  0.315 [2]     |
  0.364 [0]     |
  0.413 [233]   |
  0.462 [206]   |
  0.512 [561]   |

Latency distribution:
  10% in 0.0715 secs.
  25% in 0.0841 secs.
  50% in 0.0990 secs.
  75% in 0.1170 secs.
  90% in 0.1362 secs.
  95% in 0.1489 secs.
  99% in 0.1760 secs.
```

So in this specific case, compaction has `-5.7%` of impact in terms of `request/sec`.

File: tools/benchmark/cmd/put.go [hunk 1]
```diff
@@ -52,6 +56,9 @@ func init() {
 	putCmd.Flags().IntVar(&putTotal, "total", 10000, "Total number of put requests")
 	putCmd.Flags().IntVar(&keySpaceSize, "key-space-size", 1, "Maximum possible keys")
 	putCmd.Flags().BoolVar(&seqKeys, "sequential-keys", false, "Use sequential keys")
+	putCmd.Flags().BoolVar(&sameKey, "same-key", false, "Use one same key (useful for compaction benchmark)")
```
Comment (xiang90) [inline tools/benchmark/cmd/put.go]: we do not need same key option. if you specify key-space-size == 1, it will always put on the same key.

Comment (gyuho) [inline tools/benchmark/cmd/put.go]: Right. Will delete. Thanks.

**Comment by xiang90** [comment]
5% seems to be a pretty small number. Have you tried this multiple times? Is the result stable? The overhead is always 5%?

**Comment by xiang90** [comment]
also in the test, we can put 10k keys per second. but we only compacts 1000 revision per second. the compaction will not be able to keep up with the key put.

**Comment by gyuho** [comment]
@xiang90 I tried several times with 100k keys and they all vary between 3 ~ 7% (slower in auto-compaction side).

**Comment by xiang90** [comment]
oh. i was wrong. it is the delta not the compaction increase.

**Comment by gyuho** [comment]
Yeah compaction is 10k as fixed, I think. 
Btw. just deleted `same-key` flag.

File: tools/benchmark/cmd/put.go [hunk 2]
```diff
@@ -90,6 +96,21 @@ func putFunc(cmd *cobra.Command, args []string) {
...
+		go func() {
+			for {
+				time.Sleep(compactInterval)
+				st := time.Now()
+				err := compactKV(clients)
+				var errStr string
+				if err != nil {
+					errStr = err.Error()
+				}
+				results <- result{errStr: errStr, duration: time.Since(st)}
```
Comment (xiang90) [inline tools/benchmark/cmd/put.go]: we should not send compaction result to the put result collection.

File: tools/benchmark/cmd/put.go [hunk 3]
```diff
@@ -90,6 +96,21 @@ func putFunc(cmd *cobra.Command, args []string) {
...
 
+	if compactInterval > 0 {
+		go func() {
+			for {
+				time.Sleep(compactInterval)
+				st := time.Now()
+				err := compactKV(clients)
+				var errStr string
+				if err != nil {
+					errStr = err.Error()
```
Comment (xiang90) [inline tools/benchmark/cmd/put.go]: we probably should just panic if compaction fails.

Comment (gyuho) [inline tools/benchmark/cmd/put.go]: Sure. Will fix.

File: tools/benchmark/cmd/put.go (hunk 2 above)
Comment (gyuho) [inline tools/benchmark/cmd/put.go]: Ok will ignore the errors. Haven't got any in my test runs.

Comment (gyuho) [inline tools/benchmark/cmd/put.go]: *will panic

**Comment by mitake** [comment]
@xiang90 sure, I'll try it tomorrow. I cannot allocate a time today for a while. Could you wait for a while?

**Comment by gyuho** [comment]
@mitake Take your time! This is not urgent. Thanks!

**Comment by mitake** [comment]
@gyuho I see, thanks :)

**Comment by mitake** [comment]
@gyuho @xiang90 I tried to execute the new flags related to compaction but I got the below error:
```
$ ./benchmark --conns=10 --clients=1000 put --total=1000000 --compact-index-delta=1000 --compact-interval=0.5s
6200 / 1000000 Bom                                                                                                                                                                                ! 0.62 % 2m8spanic: rpc error: code = 11 desc = "etcdserver: storage: required revision has been compacted"

goroutine 1141 [running]:
panic(0xa71180, 0xc82088a6c0)
        /usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/coreos/etcd/tools/benchmark/cmd.compactKV(0xc8202ec000, 0x3e8, 0x3e8)
        /home/mitake/gopath/src/github.com/coreos/etcd/tools/benchmark/cmd/put.go:150 +0x389
github.com/coreos/etcd/tools/benchmark/cmd.putFunc.func2(0xc8202ec000, 0x3e8, 0x3e8)
        /home/mitake/gopath/src/github.com/coreos/etcd/tools/benchmark/cmd/put.go:103 +0x45
created by github.com/coreos/etcd/tools/benchmark/cmd.putFunc
        /home/mitake/gopath/src/github.com/coreos/etcd/tools/benchmark/cmd/put.go:105 +0x526
```

I tried `compact` branch of @gyuho 's repository. Could you take a look, @gyuho ?

**Comment by mitake** [comment]
@gyuho BTW I'm using V3DemoProcfile on single machine (of course I should test on a cluster later). Should I change a way of booting etcd servers?

**Comment by gyuho** [comment]
@mitake What was the error message that you got from `Compact`?

**Comment by gyuho** [comment]
Oh I see

```
"etcdserver: storage: required revision has been compacted"
```

**Comment by gyuho** [comment]
@mitake  So this happens when:

1. compact at revision 100
2. put... current revision 1090
3. compact at revision 90 (1090-1000)
4. 90 < 100, so returns error

It means during the compact interval, it didn't proceed fast enough to cover the compact delta.
I will cache the previous compact revision and compare two revs to prevent this.

**Comment by mitake** [comment]
@gyuho I see. I'll wait for your update.

**Comment by gyuho** [comment]
@mitake Currently there's no way to prevent it other than just print out the logs.
Can you try this again? And I recommend to pass smaller `compact-index-delta`
if you keep seeing errors. (I changed it from panic to error with log.Println)

Please let me know how it goes.

Thanks.

File: tools/benchmark/cmd/put.go [hunk 4]
```diff
@@ -113,3 +128,35 @@ func doPut(ctx context.Context, client v3.KV, requests <-chan v3.Op) {
...
+	for _, c := range clients {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		err := c.KV.Compact(ctx, revToCompact)
+		cancel()
+		// TODO: get compact revision from KV and automatically handle
+		// the case when revToCompact < compact revision.
+		if err != nil {
+			log.Printf("error (%v) when current revision %d, revision to compact %d", err, curRev, revToCompact)
+		}
+	}
```
Comment (mitake) [inline tools/benchmark/cmd/put.go]: How about simply break in successful case?

**Comment by mitake** [comment]
@gyuho I still got errors like below even with `compact-interval-delta == 1`:

```
602 / 1000000 Bm                                                                                                                                                                                 ! 0.06 % 22m9s2016-03-04 16:22:03.875702 I | error (rpc error: code = 11 desc = "etcdserver: storage: required revision has been compacted") when current revision 112404, revision to compact 112304
```

How about simply breaking the compaction request loop after success? Current loop issues compaction requests to single revision so it is causing the errors. This change solved the error: https://github.com/mitake/etcd/commit/55c27a42e0feb9dfcf29b17b3ec789cacb0f83c7

**Comment by mitake** [comment]
BTW, should compaction request to already compacted revisions be treated as an error? I'm still not sure but how about finish without error such a request simply?

**Comment by gyuho** [comment]
@mitake You are right. Compact goes through raft so we need to break after success. Just pushed fix. Can you try again? You can ignore the logs if you get only a few. Try to minimize the errors, so that we can maximize the compaction runs during benchmark.

Thanks.

**Comment by mitake** [comment]
This is my result. I tried 3 times for w/ and w/o compaction. Options are same to @gyuho 's benchmarking (without `--same-key`):

# w/o compaction
- 1st
```
Summary:
  Total:        159.0316 secs.
  Slowest:      0.8735 secs.
  Fastest:      0.0474 secs.
  Average:      0.1583 secs.
  Stddev:       0.1180 secs.
  Requests/sec: 6288.0583

Response time histogram:
  0.047 [1]     |
  0.130 [686410]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.213 [202784]        |∎∎∎∎∎∎∎∎∎∎∎
  0.295 [26722] |∎
  0.378 [19840] |∎
  0.460 [13546] |
  0.543 [10299] |
  0.626 [19852] |∎
  0.708 [13910] |
  0.791 [5962]  |
  0.873 [674]   |

Latency distribution:
  10% in 0.1063 secs.
  25% in 0.1128 secs.
  50% in 0.1222 secs.
  75% in 0.1344 secs.
  90% in 0.2510 secs.
  95% in 0.4676 secs.
  99% in 0.6808 secs.
```

- 2nd
```
Summary:
  Total:        175.4088 secs.
  Slowest:      1.0111 secs.
  Fastest:      0.0369 secs.
  Average:      0.1747 secs.
  Stddev:       0.1445 secs.
  Requests/sec: 5700.9696

Response time histogram:
  0.037 [1]     |
  0.134 [722481]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.232 [135684]        |∎∎∎∎∎∎∎
  0.329 [31727] |∎
  0.427 [25165] |∎
  0.524 [21161] |∎
  0.621 [25696] |∎
  0.719 [26071] |∎
  0.816 [9719]  |
  0.914 [2143]  |
  1.011 [152]   |

Latency distribution:
  10% in 0.1061 secs.
  25% in 0.1133 secs.
  50% in 0.1227 secs.
  75% in 0.1370 secs.
  90% in 0.3632 secs.
  95% in 0.5907 secs.
  99% in 0.7299 secs.
```

- 3rd
```
Summary:
  Total:        197.2390 secs.
  Slowest:      0.9406 secs.
  Fastest:      0.0395 secs.
  Average:      0.1966 secs.
  Stddev:       0.1683 secs.
  Requests/sec: 5069.9922

Response time histogram:
  0.039 [1]     |
  0.130 [599415]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.220 [210293]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.310 [28798] |∎
  0.400 [30555] |∎∎
  0.490 [23010] |∎
  0.580 [30485] |∎∎
  0.670 [45866] |∎∎∎
  0.760 [22554] |∎
  0.850 [8181]  |
  0.941 [842]   |

Latency distribution:
  10% in 0.1067 secs.
  25% in 0.1145 secs.
  50% in 0.1246 secs.
  75% in 0.1466 secs.
  90% in 0.5184 secs.
  95% in 0.6357 secs.
  99% in 0.7554 secs.
```

# w/ compaction

- 1st
```
Summary:
  Total:        164.0097 secs.
  Slowest:      1.1610 secs.
  Fastest:      0.0395 secs.
  Average:      0.1633 secs.
  Stddev:       0.1240 secs.
  Requests/sec: 6097.2010

Response time histogram:
  0.039 [1]     |
  0.152 [814208]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.264 [108618]        |∎∎∎∎∎
  0.376 [23988] |∎
  0.488 [14311] |
  0.600 [11893] |
  0.712 [6655]  |
  0.825 [10380] |
  0.937 [6582]  |
  1.049 [3350]  |
  1.161 [14]    |

Latency distribution:
  10% in 0.1130 secs.
  25% in 0.1192 secs.
  50% in 0.1301 secs.
  75% in 0.1449 secs.
  90% in 0.1911 secs.
  95% in 0.4162 secs.
  99% in 0.8242 secs.
```

- 2nd
```
Summary:
  Total:        165.1095 secs.
  Slowest:      1.0715 secs.
  Fastest:      0.0533 secs.
  Average:      0.1644 secs.
  Stddev:       0.1224 secs.
  Requests/sec: 6056.5861

Response time histogram:
  0.053 [1]     |
  0.155 [837092]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.257 [76034] |∎∎∎
  0.359 [24774] |∎
  0.461 [15710] |
  0.562 [13111] |
  0.664 [11530] |
  0.766 [11565] |
  0.868 [7308]  |
  0.970 [2378]  |
  1.071 [497]   |

Latency distribution:
  10% in 0.1130 secs.
  25% in 0.1188 secs.
  50% in 0.1297 secs.
  75% in 0.1446 secs.
  90% in 0.2147 secs.
  95% in 0.4395 secs.
  99% in 0.7675 secs.
```

- 3rd
```
Summary:
  Total:        162.0420 secs.
  Slowest:      1.0151 secs.
  Fastest:      0.0537 secs.
  Average:      0.1614 secs.
  Stddev:       0.1204 secs.
  Requests/sec: 6171.2405

Response time histogram:
  0.054 [1]     |
  0.150 [812465]        |∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  0.246 [105385]        |∎∎∎∎∎
  0.342 [22751] |∎
  0.438 [14110] |
  0.534 [9127]  |
  0.631 [9789]  |
  0.727 [8696]  |
  0.823 [13414] |
  0.919 [3352]  |
  1.015 [910]   |

Latency distribution:
  10% in 0.1123 secs.
  25% in 0.1184 secs.
  50% in 0.1285 secs.
  75% in 0.1429 secs.
  90% in 0.1843 secs.
  95% in 0.4035 secs.
  99% in 0.7798 secs.
```

The worst latencies could be seen w/o compaction benchmark. Maybe `--key-space-size` would change the result.

**Comment by gyuho** [comment]
@mitake Thanks a lot. I tried with `--key-space-size` flag so that we can force to write same keys. Here are my results:

```
benchmark --conns=10 --clients=1000 \
	put --total=1000000 --key-space-size=1

Summary:
  Total:        98.2062 secs.
  Slowest:      0.2805 secs.
  Fastest:      0.0227 secs.
  Average:      0.0981 secs.
  Stddev:       0.0275 secs.
  Requests/sec: 10182.6544

---

benchmark --conns=10 --clients=1000 \
	put --total=1000000  --key-space-size=1 --compact-index-delta=1000 --compact-interval=0.5s

Summary:
  Total:        106.0674 secs.
  Slowest:      0.5915 secs.
  Fastest:      0.0232 secs.
  Average:      0.1060 secs.
  Stddev:       0.0305 secs.
  Requests/sec: 9427.9705   <=============  -7.4%
```

```
benchmark --conns=10 --clients=1000 \
	put --total=1000000 --key-space-size=30

Summary:
  Total:        97.9946 secs.
  Slowest:      0.2404 secs.
  Fastest:      0.0179 secs.
  Average:      0.0979 secs.
  Stddev:       0.0259 secs.
  Requests/sec: 10204.6425

---

benchmark --conns=10 --clients=1000 \
	put --total=1000000  --key-space-size=30 --compact-index-delta=1000 --compact-interval=0.5s

Summary:
  Total:        105.8640 secs.
  Slowest:      0.3079 secs.
  Fastest:      0.0212 secs.
  Average:      0.1058 secs.
  Stddev:       0.0293 secs.
  Requests/sec: 9446.0834   <=============  -7.4%
```

So users can expect 5 ~ 7% performance impact when it compacts every half-second.

If you run with 30-second, the performance impact gets smaller:

```
benchmark --conns=10 --clients=1000 \
	put --total=1000000  --key-space-size=30 --compact-index-delta=1000 --compact-interval=30s


Summary:
  Total:        103.8649 secs.
  Slowest:      2.1752 secs.
  Fastest:      0.0126 secs.
  Average:      0.1038 secs.
  Stddev:       0.0686 secs.
  Requests/sec: 9627.8907   <=============  -5.6%
```


/cc @xiang90

**Comment by mitake** [comment]
@gyuho thanks for sharing. Your result seems to be matching with my experience: https://github.com/coreos/etcd/pull/4642

**Comment by gyuho** [comment]
@mitake @xiang90 Ignore my previous comment. I did the math wrong.
Mitake's code errors because the previous code didn't handle the loop correctly.
We just need to send compact request to one node, and let raft replicate and apply
compaction, and exit the for-loop. Previous code tries to compact to all 3-node and
it panics when it runs the duplicate compact call with the same compact revision.


Code fixed.

Thanks.

**Comment by xiang90** [comment]
LGTM

**Comment by xiang90** [comment]
LGTM

**Comment by gyuho** [comment]
Merging first, and will share more results through google docs.
Thanks.

**Comment by mitake** [comment]
@gyuho Which comment should I ignore? I'm not sure what is the difference between the latest version and the version I benchmarked. Call `panic()` in compaction failure or continue the loop?

**Comment by gyuho** [comment]
@mitake I was talking about https://github.com/coreos/etcd/pull/4638#issuecomment-192132266.

So my math was wrong.

1. error when it tries to compact with revision that is smaller than current compact revision
2. we get revision to compact by {current revision - compact delta}
3. let's say we start with current revision 10k, revision to compact 9k
4. then next round, the compact revision is 9k
5. if we try to compact at somewhere before 9k, it will error
6. but we get revision to compact always from {current revision - compact delta}, the revision to compact is every increasing
7. even if the next current revision is 10k + 1, we will compact at 9k + 1, which is still valid revision to compact

**Comment by mitake** [comment]
@gyuho I understood, thanks.

But I'm still not sure that a compaction request to already compacted revision should cause an error. Is threre a reason of the error?

**Pull request merged**